Ctrl-p rewrite with list-zipper

stites opened the pull request:
So this was intended to be an investigation into how yi-fuzzy-open works but then turned into, pretty much, a full rewrite of this package using a list zipper ([wiki](https://en.wikipedia.org/wiki/Zipper_(data_structure))) to simplify the pointer-management that's going on. I ran the brittany formatter over the yi-fuzzy-open package in the last commit. Hopefully, this makes the module easier to follow, but let me know if this is an unnecessary PR.

Changes:
+ moved inner let assignments to outer where blocks where possible.
+ `yi-core/src/Yi/Completion.hs`
  - add `subsequenceTextMatch`, which comes from a TODO.
  - there's an indentation discrepancy which slipped in. Do you have a formatting / linting guide? 
+ ` yi-fuzzy-open/src/Yi/Fuzzy/Zipper.hs`
  - add a list-zipper (which can be empty). I rewrote a simple one instead of relying on kmett's zippers-package because of the ubiquitous lens dependency in most of kmett's projects.
+ `yi-fuzzy-open/src/Yi/Fuzzy.hs` -->
  - basically, this turned into a close-to-full rewrite so I formatted things the way I usually develop.
  - I removed the import is a hack to silence redundant imports, I'm not sure warnings are a bad thing(?)
  - `FuzzyItem` from a record type to a sum-type (with FilePath stored as Text)
  - Use `Natural` when there is a requirement for non-negative integers
  - ADDED: order Buffers first -- this was because I figured buffers would be more commonly requested than files. It also makes `filterNotCommon :: [FuzzyItem] -> [FuzzyItem]` a requirement for things like the console buffer.
  - CHANGED: increment/decrement-Index is now goLeft/Right. This is more intuitive for the current yi-fuzzy-open since we only have a single status-line to move along. I might try to spend some time on a multi-line pane.
  - ADDED: `<Tab>` and `<Shift-Tab>` to navigation in the local keymap.
  - use `Word8` for binary representation. This is more compact than Int.

Note: viewing changes in github will be easier with split view, rather than the unified view.

ethercrow commented:
yi already depends on pointedlist package, could you use that as a zipper list?

ethercrow commented:
> Do you have a formatting / linting guide?

No, just be consistent within a module.

noughtmare commented on yi-core/src/Yi/Completion.hs:
```diff
@@ -56,11 +58,26 @@ infixUptoEndMatch needle haystack = case T.breakOn needle haystack of
+-- | A simple fuzzy match algorithm. Example: "abc" matches "a1b2c"
+subsequenceTextMatch :: Text -> Text -> Bool
+subsequenceTextMatch needle haystack =
+  if | T.null needle   -> True
```
The multi-way if can be avoided by just using guards:
http://learnyouahaskell.com/syntax-in-functions#guards-guards

noughtmare commented on yi-fuzzy-open/src/Yi/Fuzzy.hs:
```diff
@@ -17,55 +19,65 @@
+instance Show FuzzyItem where
+  -- TODO: make subsequenceMatch work on Text
+  show :: FuzzyItem -> String
+  show = \case
```
This case can be avoided by just using pattern matching

noughtmare commented on yi-fuzzy-open/src/Yi/Fuzzy.hs:
```diff
@@ -17,55 +19,65 @@
+    i@(BufferItem _) -> "Buffer  " <> itemAsStr i
+
+itemAsTxt :: FuzzyItem -> Text
+itemAsTxt = \case
```
Lambda case can be avoided here as well.

noughtmare commented on yi-fuzzy-open/src/Yi/Fuzzy.hs:
```diff
@@ -76,133 +88,114 @@ fuzzyOpen = fuzzyOpenWithDepth defaultDepth
-        putBufferDyn initialState
-    withEditor (renderE initialState)
+fuzzyOpenWithDepth :: Natural -> YiM ()
+fuzzyOpenWithDepth (fromIntegral->d) = do
```
I am also against using viewpatterns here, because `d` is only used once in this function, why not just put the `fromIntegral` function at the place where `d` is used?

noughtmare commented on yi-fuzzy-open/src/Yi/Fuzzy.hs:
```diff
@@ -215,21 +208,24 @@ openInNewTab = openRoutine newTabE
+      action i
+ where
+  action :: FuzzyItem -> YiM ()
+  action = \case
```
This is also an avoidable lambdacase.

noughtmare commented:
I have added some comments about (in my opinion) unnecessary language extensions. I try to avoid them as much as possible, because they are not beginner friendly and they are specific to GHC.

stites commented:
@ethercrow -- I've moved this PR to use pointedlist. Since pointedlist is a non-empty zipper and the fuzzy-search can be empty (when the search filters down the items), the `items` type has changed to be a `Maybe (PointedList FuzzyItem)`. I've also renamed my `goLeft` (decrementIndex) and `goRight` (incrementIndex) to `goPrevious` and `goNext`, respectively. I've left the module formatting as-is since you said I should just be consistent within the module.
 
@noughtmare -- I've removed dependencies on MultiWayIf and LambdaCase. As you can tell I'm a big fan of LambdaCase, so it was a little heartwrenching : )

There are some usability bugs that exist in this module that I'd like to tackle later (namely: rendering in a larger buffer, and how much of the left list we show), but I don't think it belongs in this PR.

Also, I added "messages" to the filtered out common buffer names

ethercrow commented on yi-fuzzy-open/src/Yi/Fuzzy.hs:
```diff
@@ -76,133 +89,134 @@ fuzzyOpen = fuzzyOpenWithDepth defaultDepth
+  filterNotCommon = filter ((\n -> not (n == "console" || n == "messages")) . itemAsTxt)
+
 
 -- shamelessly stolen from Chapter 9 of Real World Haskell
```
I think you can delete that note now that you've pretty much rewritten it.

stites replied:
👍

ethercrow commented:
Thanks!
